In [26]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms,models
from torch.utils.data import DataLoader, random_split
from tqdm import tqdm
import kagglehub
from pathlib import Path
import os
from torch.utils.data import random_split
import torch.nn.functional as F
from PIL import Image
from PIL import UnidentifiedImageError
import neptune


In [27]:
path = kagglehub.dataset_download("shaunthesheep/microsoft-catsvsdogs-dataset")
print(path)
data_dir = Path(path) / "PetImages"
print(data_dir)

/kaggle/input/microsoft-catsvsdogs-dataset
/kaggle/input/microsoft-catsvsdogs-dataset/PetImages


In [28]:
class SafeImageFolder(datasets.ImageFolder):
    def __getitem__(self, index):
        try:
            return super().__getitem__(index)
        except (OSError, UnidentifiedImageError, ValueError) as e:
            print(f"Пропущен битый файл: {self.samples[index][0]} ({e})")
            
            new_index = (index + 1) % len(self.samples)
            return self.__getitem__(new_index)

In [29]:
transform = transforms.Compose([
    transforms.Resize((128, 128)),
    transforms.ToTensor(),
])

dataset=SafeImageFolder(root=data_dir,transform=transform)

train_size = int(0.8 * len(dataset))
test_size = len(dataset) - train_size
train_dataset, test_dataset = random_split(dataset, [train_size, test_size])



train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True,num_workers=4)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False,num_workers=4)
num_classes=2


In [30]:
class LeNet5(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1=nn.Conv2d(3,6,kernel_size=5,stride=1,padding=2)
        self.pool1=nn.AvgPool2d(2,2)
        self.conv2=nn.Conv2d(6,16,kernel_size=5)
        self.pool2=nn.AvgPool2d(2,2)
        self.conv3=nn.Conv2d(16,120,kernel_size=5)
        
        self.fc1 = None  
        self.fc2 = nn.Linear(84, num_classes)


    def forward_0(self, x):
        x = F.relu(self.conv1(x))
        x = self.pool1(x)
        x = F.relu(self.conv2(x))
        x = self.pool2(x)
        x = F.relu(self.conv3(x))
        return x 

    def forward(self,x):
        x=self.forward_0(x)
        if self.fc1 is None:
            flattened_dim = x.shape[1] * x.shape[2] * x.shape[3]
            self.fc1 = nn.Linear(flattened_dim, 84).to(x.device)
        
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x

In [31]:
def evaluate(model,dataloader,device):
    model.eval()
    correct=0
    total=0
    running_loss=0.0
    with torch.no_grad():
        for images,labels in dataloader:
            images = images.to(device)
            labels = labels.to(device)
            outputs=model(images)
            loss=criterion(outputs,labels)
            running_loss += loss.item() * images.size(0)
            preds = outputs.argmax(dim=1)
            correct += (preds == labels).sum().item()
            total += labels.size(0)
    avg_loss = running_loss / total
    accuracy = correct / total
    model.train()
    return avg_loss, accuracy

In [32]:

run = neptune.init_run(
    project="kulikovka/AlexLenNets",
    api_token="eyJhcGlfYWRkcmVzcyI6Imh0dHBzOi8vYXBwLm5lcHR1bmUuYWkiLCJhcGlfdXJsIjoiaHR0cHM6Ly9hcHAubmVwdHVuZS5haSIsImFwaV9rZXkiOiJiNjZjNWE3OS1jMzRlLTQyZjAtYmFiNi04ZGU1ZjY0MDAxNzgifQ=="
)


device = torch.device("cuda")
model=LeNet5().to(device)
criterion=nn.CrossEntropyLoss()
optimazer=optim.Adam(model.parameters(),lr=0.001)
num_epochs=15

train_losses=[]
train_accs=[]
test_losses=[]
test_accs=[]

run['parameters'] = {
    "epochs": num_epochs,
    "device": str(device),
    "train_size": len(train_dataset),
    "test_size": len(test_dataset)
}



[neptune] [info   ] Neptune initialized. Open in the app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-10


In [33]:

for epoch in range(1, num_epochs + 1):
    model.train()
    epoch_loss = 0.0
    correct = 0
    total = 0

    pbar = tqdm(train_loader, desc=f"Epoch {epoch}/{num_epochs}")

    for images, labels in pbar:
        images = images.to(device)
        labels = labels.to(device)
        optimazer.zero_grad()           
        outputs = model(images)         
        loss = criterion(outputs, labels)  
        loss.backward()                 
        optimazer.step()                

        epoch_loss += loss.item() * images.size(0)
        preds = outputs.argmax(dim=1)
        correct += (preds == labels).sum().item()
        total += labels.size(0)

        pbar.set_postfix({"batch_loss": loss.item(), "acc": f"{100 * correct / total}%"})

    train_loss = epoch_loss / len(train_dataset)
    train_acc = correct / total
    test_loss, test_acc = evaluate(model, test_loader,device)

    print(f"Epoch {epoch}: Train loss={train_loss}, Train acc={train_acc}, "
          f"Test loss={test_loss}, Test acc={test_acc}")
    run["train/loss"].append(train_loss)
    run["train/accuracy"].append(train_acc)
    run["test/loss"].append(test_loss)
    run["test/accuracy"].append(test_acc)
    run["epoch"].append(epoch)
run.stop()

Epoch 1/15:   0%|          | 1/313 [00:01<09:25,  1.81s/it, batch_loss=0.692, acc=53.125%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 1/15: 100%|██████████| 313/313 [00:27<00:00, 11.48it/s, batch_loss=0.672, acc=56.57%]             
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 1: Train loss=0.6779731781005859, Train acc=0.5657, Test loss=0.652942924118042, Test acc=0.6178


Epoch 2/15:  62%|██████▏   | 193/313 [00:17<00:09, 12.48it/s, batch_loss=0.636, acc=62.28950777202073%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 2/15: 100%|██████████| 313/313 [00:27<00:00, 11.42it/s, batch_loss=0.676, acc=62.58%]             
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 2: Train loss=0.6443394677162171, Train acc=0.6258, Test loss=0.6352069123268127, Test acc=0.6382


Epoch 3/15:  65%|██████▌   | 205/313 [00:18<00:08, 12.04it/s, batch_loss=0.571, acc=66.89453125%]      

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 3/15: 100%|██████████| 313/313 [00:27<00:00, 11.38it/s, batch_loss=0.608, acc=67.375%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 3: Train loss=0.6052867392539978, Train acc=0.67375, Test loss=0.5846970951080322, Test acc=0.689


Epoch 4/15:  69%|██████▉   | 216/313 [00:19<00:07, 13.76it/s, batch_loss=0.569, acc=69.87123842592592%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 4/15: 100%|██████████| 313/313 [00:27<00:00, 11.45it/s, batch_loss=0.588, acc=70.035%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 4: Train loss=0.5791666253089904, Train acc=0.70035, Test loss=0.5723871466636657, Test acc=0.7072


Epoch 5/15:  48%|████▊     | 150/313 [00:14<00:14, 11.46it/s, batch_loss=0.574, acc=70.9498355263158%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 5/15: 100%|██████████| 313/313 [00:27<00:00, 11.40it/s, batch_loss=0.558, acc=70.87%]            
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 5: Train loss=0.5640148385047913, Train acc=0.7087, Test loss=0.5957524838447571, Test acc=0.6776


Epoch 6/15:  26%|██▌       | 81/313 [00:08<00:19, 11.88it/s, batch_loss=0.494, acc=72.50381097560975%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 6/15: 100%|██████████| 313/313 [00:27<00:00, 11.48it/s, batch_loss=0.69, acc=71.84%]             
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 6: Train loss=0.550806381034851, Train acc=0.7184, Test loss=0.5648667139053345, Test acc=0.7068


Epoch 7/15:  69%|██████▉   | 217/313 [00:19<00:08, 11.80it/s, batch_loss=0.48, acc=72.57344470046083%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 7/15: 100%|██████████| 313/313 [00:27<00:00, 11.57it/s, batch_loss=0.66, acc=73.08%]             
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 7: Train loss=0.5352104139328003, Train acc=0.7308, Test loss=0.5480288083076477, Test acc=0.7238


Epoch 8/15:  25%|██▌       | 79/313 [00:08<00:18, 12.61it/s, batch_loss=0.527, acc=74.6484375%]       

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 8/15: 100%|██████████| 313/313 [00:27<00:00, 11.31it/s, batch_loss=0.465, acc=74.485%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 8: Train loss=0.5189290283679963, Train acc=0.74485, Test loss=0.5546228767871857, Test acc=0.716


Epoch 9/15:  30%|███       | 94/313 [00:09<00:18, 11.66it/s, batch_loss=0.574, acc=76.15559895833333%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 9/15: 100%|██████████| 313/313 [00:27<00:00, 11.43it/s, batch_loss=0.41, acc=74.875%]            
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 9: Train loss=0.5129279798030854, Train acc=0.74875, Test loss=0.5502178789615632, Test acc=0.7246


Epoch 10/15:  72%|███████▏  | 225/313 [00:20<00:07, 11.16it/s, batch_loss=0.495, acc=75.82236842105263%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 10/15: 100%|██████████| 313/313 [00:27<00:00, 11.27it/s, batch_loss=0.429, acc=75.98%]            
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 10: Train loss=0.4995955293655395, Train acc=0.7598, Test loss=0.5388641901493072, Test acc=0.7242


Epoch 11/15:  30%|███       | 94/313 [00:09<00:18, 11.95it/s, batch_loss=0.479, acc=76.87174479166667%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)

Epoch 11/15:  31%|███       | 96/313 [00:09<00:19, 11.06it/s, batch_loss=0.479, acc=76.87174479166667%]

Epoch 11/15: 100%|██████████| 313/313 [00:27<00:00, 11.34it/s, batch_loss=0.512, acc=76.545%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 11: Train loss=0.4876128788948059, Train acc=0.76545, Test loss=0.5549858328819275, Test acc=0.7216


Epoch 12/15:  32%|███▏      | 100/313 [00:09<00:15, 14.04it/s, batch_loss=0.436, acc=77.140625%]       

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 12/15: 100%|██████████| 313/313 [00:26<00:00, 11.65it/s, batch_loss=0.478, acc=77.385%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 12: Train loss=0.47517421469688415, Train acc=0.77385, Test loss=0.5292804244995117, Test acc=0.741


Epoch 13/15:  16%|█▋        | 51/313 [00:06<00:21, 11.93it/s, batch_loss=0.504, acc=78.6358173076923%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 13/15: 100%|██████████| 313/313 [00:27<00:00, 11.23it/s, batch_loss=0.502, acc=78.22%]            
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 13: Train loss=0.4644625379562378, Train acc=0.7822, Test loss=0.5281127941608429, Test acc=0.739


Epoch 14/15:  63%|██████▎   | 198/313 [00:17<00:09, 11.99it/s, batch_loss=0.436, acc=78.88257575757575%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 14/15: 100%|██████████| 313/313 [00:27<00:00, 11.54it/s, batch_loss=0.419, acc=78.865%]           
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 14: Train loss=0.45385634846687317, Train acc=0.78865, Test loss=0.5287555519580841, Test acc=0.7382


Epoch 15/15:  51%|█████▏    | 161/313 [00:15<00:12, 11.71it/s, batch_loss=0.379, acc=79.7360248447205%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 15/15: 100%|██████████| 313/313 [00:27<00:00, 11.35it/s, batch_loss=0.45, acc=79.505%]            
/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)
Epoch 15: Train loss=0.4429778965950012, Train acc=0.79505, Test loss=0.5385178909301758, Test acc=0.7368
[neptune] [info   ] Shutting down background jobs, please wait a moment...
[neptune] [info   ] Done!
[neptune] [info   ] Waiting for the remaining 5 operations to synchronize with Neptune. Do not kill this process.
[neptune] [info   ] All 5 operations synced, thanks for waiting!
[neptune] [info   ] Explore the metadata in the Neptune app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-10/metadata


In [79]:
class AlexNet(nn.Module):
    def __init__(self, num_classes=2):
        super().__init__()
        self.conv_layers = nn.Sequential(
            nn.Conv2d(3, 96, kernel_size=11, stride=4, padding=2),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2),

            nn.Conv2d(96, 256, kernel_size=5, padding=2),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2),

            nn.Conv2d(256, 384, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),

            nn.Conv2d(384, 384, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),

            nn.Conv2d(384, 256, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2)
        )

        self.fc_layers = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(256 * 6 * 6, 4096),  
            nn.ReLU(inplace=True),
            nn.Dropout(0.5),
            nn.Linear(4096, 4096),
            nn.ReLU(inplace=True),
            nn.Linear(4096, num_classes)
        )

    def forward(self, x):
        x = self.conv_layers(x)
        x = torch.flatten(x, 1)
        x = self.fc_layers(x)
        return x

In [80]:
run = neptune.init_run(
    project="kulikovka/AlexLenNets",
    api_token="eyJhcGlfYWRkcmVzcyI6Imh0dHBzOi8vYXBwLm5lcHR1bmUuYWkiLCJhcGlfdXJsIjoiaHR0cHM6Ly9hcHAubmVwdHVuZS5haSIsImFwaV9rZXkiOiJiNjZjNWE3OS1jMzRlLTQyZjAtYmFiNi04ZGU1ZjY0MDAxNzgifQ=="
)


# ---------------- Training ----------------
model = AlexNet(num_classes=num_classes).to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9, weight_decay=5e-4)
num_epochs = 15


run['parameters'] = {
    "epochs": num_epochs,
    "device": str(device),
    "train_size": len(train_dataset),
    "test_size": len(test_dataset),
    "lr": 0.01
}


[neptune] [info   ] Neptune initialized. Open in the app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-24


In [81]:

# ---------------- Neptune ----------------

for epoch in range(1, num_epochs + 1):
    model.train()
    epoch_loss, correct, total = 0.0, 0, 0
    pbar = tqdm(train_loader, desc=f"Epoch {epoch}/{num_epochs}")
    
    for images, labels in pbar:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        
        epoch_loss += loss.item() * images.size(0)
        preds = outputs.argmax(dim=1)
        correct += (preds == labels).sum().item()
        total += labels.size(0)
        
        pbar.set_postfix({"batch_loss": loss.item(), "acc": f"{100*correct/total:.2f}%"})
    
    train_loss = epoch_loss / len(train_dataset)
    train_acc = correct / total
    test_loss, test_acc = evaluate(model, test_loader, device)
    
    print(f"Epoch {epoch}: Train loss={train_loss:.4f}, Train acc={train_acc:.4f}, "
          f"Test loss={test_loss:.4f}, Test acc={test_acc:.4f}")
    
    run["train/loss"].append(train_loss)
    run["train/accuracy"].append(train_acc)
    run["test/loss"].append(test_loss)
    run["test/accuracy"].append(test_acc)
    run["epoch"].append(epoch)

run.stop()


Epoch 1/15:  19%|█▉        | 61/313 [00:12<00:37,  6.65it/s, batch_loss=0.693, acc=51.31%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 1/15:  46%|████▌     | 143/313 [00:23<00:23,  7.37it/s, batch_loss=0.695, acc=52.90%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 1/15:  78%|███████▊  | 243/313 [00:38<00:09,  7.37it/s, batch_loss=0.681, acc=55.06%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 1/15: 100%|██████████| 313/313 [00:48<00:00,  6.44it/s, batch_loss=0.593, acc=55.88%]


Epoch 1: Train loss=0.6810, Train acc=0.5588, Test loss=0.6978, Test acc=0.5896


Epoch 2/15:  13%|█▎        | 40/313 [00:09<00:33,  8.12it/s, batch_loss=0.615, acc=59.73%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 2/15:  34%|███▎      | 105/313 [00:18<00:31,  6.61it/s, batch_loss=0.6, acc=61.32%]  

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 2/15:  80%|███████▉  | 249/313 [00:39<00:09,  6.44it/s, batch_loss=0.781, acc=62.87%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 2/15: 100%|██████████| 313/313 [00:48<00:00,  6.52it/s, batch_loss=0.601, acc=63.84%]


Epoch 2: Train loss=0.6335, Train acc=0.6384, Test loss=0.5935, Test acc=0.6868


Epoch 3/15:  47%|████▋     | 146/313 [00:24<00:24,  6.87it/s, batch_loss=0.611, acc=69.75%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 3/15:  54%|█████▎    | 168/313 [00:27<00:18,  7.95it/s, batch_loss=0.636, acc=70.15%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 3/15:  60%|██████    | 189/313 [00:30<00:19,  6.45it/s, batch_loss=0.708, acc=70.17%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 3/15: 100%|██████████| 313/313 [00:47<00:00,  6.55it/s, batch_loss=0.505, acc=71.92%]


Epoch 3: Train loss=0.5513, Train acc=0.7191, Test loss=0.4933, Test acc=0.7574


Epoch 4/15:  24%|██▍       | 76/313 [00:16<00:32,  7.38it/s, batch_loss=0.395, acc=74.61%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 4/15:  51%|█████     | 159/313 [00:27<00:20,  7.57it/s, batch_loss=0.423, acc=75.89%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 4/15:  76%|███████▌  | 237/313 [00:38<00:11,  6.34it/s, batch_loss=0.477, acc=76.89%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 4/15: 100%|██████████| 313/313 [00:49<00:00,  6.32it/s, batch_loss=0.406, acc=77.33%]


Epoch 4: Train loss=0.4772, Train acc=0.7733, Test loss=0.4870, Test acc=0.7532


Epoch 5/15:   8%|▊         | 26/313 [00:07<00:39,  7.27it/s, batch_loss=0.391, acc=81.01%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 5/15:  37%|███▋      | 116/313 [00:19<00:24,  7.88it/s, batch_loss=0.614, acc=81.53%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 5/15:  63%|██████▎   | 196/313 [00:31<00:14,  7.95it/s, batch_loss=0.472, acc=81.60%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 5/15: 100%|██████████| 313/313 [00:47<00:00,  6.54it/s, batch_loss=0.388, acc=82.02%]


Epoch 5: Train loss=0.3999, Train acc=0.8202, Test loss=0.3560, Test acc=0.8432


Epoch 6/15:  46%|████▋     | 145/313 [00:24<00:26,  6.36it/s, batch_loss=0.372, acc=83.74%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 6/15:  54%|█████▍    | 169/313 [00:27<00:21,  6.65it/s, batch_loss=0.464, acc=83.84%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 6/15:  81%|████████  | 254/313 [00:39<00:08,  6.78it/s, batch_loss=0.415, acc=84.44%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 6/15: 100%|██████████| 313/313 [00:48<00:00,  6.50it/s, batch_loss=0.24, acc=84.38%] 


Epoch 6: Train loss=0.3507, Train acc=0.8438, Test loss=0.3450, Test acc=0.8506


Epoch 7/15:  22%|██▏       | 69/313 [00:13<00:37,  6.49it/s, batch_loss=0.294, acc=86.00%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)

Epoch 7/15:  23%|██▎       | 71/313 [00:13<00:33,  7.29it/s, batch_loss=0.441, acc=85.94%]

Epoch 7/15:  25%|██▌       | 79/313 [00:14<00:33,  6.97it/s, batch_loss=0.401, acc=85.74%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 7/15:  77%|███████▋  | 241/313 [00:37<00:10,  6.57it/s, batch_loss=0.253, acc=86.42%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 7/15: 100%|██████████| 313/313 [00:47<00:00,  6.53it/s, batch_loss=0.436, acc=86.72%]


Epoch 7: Train loss=0.3087, Train acc=0.8672, Test loss=0.3117, Test acc=0.8634


Epoch 8/15:  24%|██▍       | 75/313 [00:14<00:34,  6.95it/s, batch_loss=0.541, acc=86.70%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 8/15:  29%|██▉       | 91/313 [00:16<00:29,  7.45it/s, batch_loss=0.217, acc=87.06%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 8/15:  93%|█████████▎| 292/313 [00:45<00:02,  8.04it/s, batch_loss=0.296, acc=87.71%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 8/15: 100%|██████████| 313/313 [00:48<00:00,  6.51it/s, batch_loss=0.193, acc=87.82%]


Epoch 8: Train loss=0.2823, Train acc=0.8782, Test loss=0.2928, Test acc=0.8728


Epoch 9/15:   4%|▎         | 11/313 [00:05<01:03,  4.79it/s, batch_loss=0.212, acc=89.91%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 9/15:  27%|██▋       | 85/313 [00:15<00:34,  6.60it/s, batch_loss=0.286, acc=89.69%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 9/15:  62%|██████▏   | 194/313 [00:31<00:17,  6.98it/s, batch_loss=0.295, acc=89.19%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 9/15: 100%|██████████| 313/313 [00:48<00:00,  6.49it/s, batch_loss=0.104, acc=89.44%] 


Epoch 9: Train loss=0.2474, Train acc=0.8944, Test loss=0.2736, Test acc=0.8856


Epoch 10/15:  44%|████▍     | 137/313 [00:23<00:26,  6.54it/s, batch_loss=0.203, acc=89.49%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 10/15:  50%|████▉     | 155/313 [00:25<00:21,  7.43it/s, batch_loss=0.257, acc=89.47%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 10/15:  53%|█████▎    | 167/313 [00:27<00:19,  7.36it/s, batch_loss=0.238, acc=89.49%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 10/15: 100%|██████████| 313/313 [00:48<00:00,  6.48it/s, batch_loss=0.466, acc=89.90%] 


Epoch 10: Train loss=0.2354, Train acc=0.8990, Test loss=0.2794, Test acc=0.8820


Epoch 11/15:  41%|████      | 129/313 [00:22<00:31,  5.84it/s, batch_loss=0.126, acc=90.99%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 11/15:  76%|███████▋  | 239/313 [00:37<00:10,  7.29it/s, batch_loss=0.218, acc=91.13%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 11/15:  89%|████████▉ | 279/313 [00:43<00:04,  7.49it/s, batch_loss=0.118, acc=91.07%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 11/15: 100%|██████████| 313/313 [00:48<00:00,  6.50it/s, batch_loss=0.41, acc=91.02%] 


Epoch 11: Train loss=0.2162, Train acc=0.9102, Test loss=0.2511, Test acc=0.8998


Epoch 12/15:  11%|█         | 34/313 [00:08<00:39,  7.12it/s, batch_loss=0.224, acc=92.14%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 12/15:  28%|██▊       | 89/313 [00:16<00:34,  6.45it/s, batch_loss=0.197, acc=91.89%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 12/15:  31%|███       | 97/313 [00:17<00:31,  6.86it/s, batch_loss=0.288, acc=91.77%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 12/15: 100%|██████████| 313/313 [00:47<00:00,  6.56it/s, batch_loss=0.201, acc=91.59%] 


Epoch 12: Train loss=0.2033, Train acc=0.9160, Test loss=0.2196, Test acc=0.9128


Epoch 13/15:   7%|▋         | 23/313 [00:07<00:40,  7.22it/s, batch_loss=0.16, acc=92.19%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)

Epoch 13/15:   8%|▊         | 25/313 [00:07<00:42,  6.70it/s, batch_loss=0.16, acc=92.19%]

Epoch 13/15:  29%|██▉       | 92/313 [00:16<00:27,  8.01it/s, batch_loss=0.159, acc=92.41%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 13/15:  30%|██▉       | 93/313 [00:16<00:34,  6.43it/s, batch_loss=0.211, acc=92.41%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)

Epoch 13/15:  30%|███       | 94/313 [00:16<00:31,  6.98it/s, batch_loss=0.298, acc=92.27%]

Epoch 13/15: 100%|██████████| 313/313 [00:47<00:00,  6.54it/s, batch_loss=0.273, acc=92.18%] 


Epoch 13: Train loss=0.1912, Train acc=0.9218, Test loss=0.2132, Test acc=0.9098


Epoch 14/15:   6%|▌         | 19/313 [00:06<00:44,  6.62it/s, batch_loss=0.129, acc=91.72%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 14/15:  84%|████████▎ | 262/313 [00:40<00:07,  6.78it/s, batch_loss=0.181, acc=92.64%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 14/15:  85%|████████▍ | 265/313 [00:41<00:07,  6.09it/s, batch_loss=0.136, acc=92.70%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 14/15: 100%|██████████| 313/313 [00:47<00:00,  6.56it/s, batch_loss=0.11, acc=92.69%]  


Epoch 14: Train loss=0.1789, Train acc=0.9268, Test loss=0.2218, Test acc=0.9142


Epoch 15/15:   4%|▎         | 11/313 [00:05<00:53,  5.67it/s, batch_loss=0.28, acc=93.36%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 15/15:  54%|█████▍    | 170/313 [00:27<00:20,  7.04it/s, batch_loss=0.156, acc=93.05%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 15/15:  66%|██████▌   | 207/313 [00:32<00:13,  7.61it/s, batch_loss=0.101, acc=93.06%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 15/15: 100%|██████████| 313/313 [00:47<00:00,  6.59it/s, batch_loss=0.171, acc=93.00%] 


Epoch 15: Train loss=0.1727, Train acc=0.9300, Test loss=0.2177, Test acc=0.9062
[neptune] [info   ] Shutting down background jobs, please wait a moment...
[neptune] [info   ] Done!
[neptune] [info   ] Waiting for the remaining 5 operations to synchronize with Neptune. Do not kill this process.
[neptune] [info   ] All 5 operations synced, thanks for waiting!
[neptune] [info   ] Explore the metadata in the Neptune app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-24/metadata


In [82]:
alexnet = models.alexnet(weights=models.AlexNet_Weights.IMAGENET1K_V1)



In [83]:
for param in alexnet.features.parameters():
    param.requires_grad = False

In [84]:
alexnet.classifier = nn.Sequential(
    nn.Linear(256 * 6 * 6, 4096),
    nn.ReLU(inplace=True),
    nn.Linear(4096, 1024),
    nn.ReLU(inplace=True),
    nn.Linear(1024, num_classes)
)



In [87]:
run = neptune.init_run(
    project="kulikovka/AlexLenNets",
    api_token="eyJhcGlfYWRkcmVzcyI6Imh0dHBzOi8vYXBwLm5lcHR1bmUuYWkiLCJhcGlfdXJsIjoiaHR0cHM6Ly9hcHAubmVwdHVuZS5haSIsImFwaV9rZXkiOiJiNjZjNWE3OS1jMzRlLTQyZjAtYmFiNi04ZGU1ZjY0MDAxNzgifQ=="
)

model = alexnet.to(device)
trainable_params = [p for p in model.parameters() if p.requires_grad]
optimizer = optim.Adam(trainable_params, lr=0.001)
criterion=nn.CrossEntropyLoss()

num_epochs=5

train_losses=[]
train_accs=[]
test_losses=[]
test_accs=[]

run['parameters'] = {
    "epochs": num_epochs,
    "device": str(device),
    "train_size": len(train_dataset),
    "test_size": len(test_dataset)
}

[neptune] [info   ] Neptune initialized. Open in the app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-26


In [88]:
for epoch in range(1, num_epochs + 1):
    model.train()
    epoch_loss = 0.0
    correct = 0
    total = 0

    pbar = tqdm(train_loader, desc=f"Epoch {epoch}/{num_epochs}")

    for images, labels in pbar:
        images = images.to(device)
        labels = labels.to(device)
        optimizer.zero_grad()           
        outputs = model(images)         
        loss = criterion(outputs, labels)  
        loss.backward()                 
        optimizer.step()                

        epoch_loss += loss.item() * images.size(0)
        preds = outputs.argmax(dim=1)
        correct += (preds == labels).sum().item()
        total += labels.size(0)

        pbar.set_postfix({"batch_loss": loss.item(), "acc": f"{100 * correct / total}%"})

    train_loss = epoch_loss / len(train_dataset)
    train_acc = correct / total
    test_loss, test_acc = evaluate(model, test_loader,device)

    print(f"Epoch {epoch}: Train loss={train_loss}, Train acc={train_acc}, "
          f"Test loss={test_loss}, Test acc={test_acc}")
    run["train/loss"].append(train_loss)
    run["train/accuracy"].append(train_acc)
    run["test/loss"].append(test_loss)
    run["test/accuracy"].append(test_acc)
    run["epoch"].append(epoch)

run.stop()

Epoch 1/5:   4%|▎         | 11/313 [00:05<01:09,  4.35it/s, batch_loss=0.101, acc=95.83333333333333%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 1/5:  27%|██▋       | 84/313 [00:14<00:28,  8.15it/s, batch_loss=0.0707, acc=95.72172619047619%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 1/5:  35%|███▍      | 108/313 [00:18<00:24,  8.22it/s, batch_loss=0.166, acc=95.52951388888889%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 1/5: 100%|██████████| 313/313 [00:45<00:00,  6.84it/s, batch_loss=0.133, acc=95.405%]            


Epoch 1: Train loss=0.1181787448361516, Train acc=0.95405, Test loss=0.13219477586746217, Test acc=0.9492


Epoch 2/5:   1%|          | 3/313 [00:04<05:26,  1.05s/it, batch_loss=0.107, acc=96.875%]           

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 2/5:  60%|██████    | 188/313 [00:29<00:15,  8.31it/s, batch_loss=0.081, acc=96.00232712765957%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 2/5:  96%|█████████▌| 300/313 [00:44<00:01,  8.39it/s, batch_loss=0.0434, acc=95.97395833333333%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 2/5: 100%|██████████| 313/313 [00:46<00:00,  6.78it/s, batch_loss=0.2, acc=95.93%]               


Epoch 2: Train loss=0.1049180777490139, Train acc=0.9593, Test loss=0.10441287317276, Test acc=0.9624


Epoch 3/5:  14%|█▎        | 43/313 [00:09<00:33,  8.13it/s, batch_loss=0.131, acc=96.83948863636364%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 3/5:  78%|███████▊  | 244/313 [00:36<00:08,  8.33it/s, batch_loss=0.0743, acc=96.71490778688525%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)

Epoch 3/5:  78%|███████▊  | 244/313 [00:36<00:08,  8.33it/s, batch_loss=0.196, acc=96.70280612244898%] 

Epoch 3/5:  81%|████████  | 252/313 [00:37<00:07,  8.32it/s, batch_loss=0.072, acc=96.69589920948617%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)

Epoch 3/5:  81%|████████  | 254/313 [00:37<00:08,  7.09it/s, batch_loss=0.107, acc=96.69045275590551%]

Epoch 3/5: 100%|██████████| 313/313 [00:45<00:00,  6.88it/s, batch_loss=0.0712, acc=96.515%]           


Epoch 3: Train loss=0.08724670278280973, Train acc=0.96515, Test loss=0.12800972115397452, Test acc=0.96


Epoch 4/5:   7%|▋         | 23/313 [00:06<00:38,  7.46it/s, batch_loss=0.0939, acc=97.72135416666667%]/usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 4/5:  39%|███▊      | 121/313 [00:20<00:27,  7.08it/s, batch_loss=0.0365, acc=97.14395491803279%] 

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 4/5:  53%|█████▎    | 167/313 [00:26<00:17,  8.27it/s, batch_loss=0.0543, acc=97.04241071428571%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)


Epoch 4/5: 100%|██████████| 313/313 [00:45<00:00,  6.85it/s, batch_loss=0.0971, acc=96.87%]            


Epoch 4: Train loss=0.07985325827896594, Train acc=0.9687, Test loss=0.13820140643715859, Test acc=0.952


Epoch 5/5:   0%|          | 1/313 [00:03<20:12,  3.89s/it, batch_loss=0.027, acc=97.65625%]

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Cat/666.jpg'>)

Epoch 5/5:   1%|          | 3/313 [00:04<05:28,  1.06s/it, batch_loss=0.0109, acc=98.4375%]

Epoch 5/5:  65%|██████▍   | 203/313 [00:30<00:13,  8.16it/s, batch_loss=0.0956, acc=97.38051470588235%] /usr/local/lib/python3.11/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))
Epoch 5/5:  66%|██████▌   | 207/313 [00:31<00:12,  8.41it/s, batch_loss=0.0183, acc=97.36328125%]      

Пропущен битый файл: /kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg (cannot identify image file <_io.BufferedReader name='/kaggle/input/microsoft-catsvsdogs-dataset/PetImages/Dog/11702.jpg'>)


Epoch 5/5: 100%|██████████| 313/313 [00:45<00:00,  6.89it/s, batch_loss=0.121, acc=97.23%]              


Epoch 5: Train loss=0.0740445006147027, Train acc=0.9723, Test loss=0.13322922866344453, Test acc=0.9498
[neptune] [info   ] Shutting down background jobs, please wait a moment...
[neptune] [info   ] Done!
[neptune] [info   ] Waiting for the remaining 5 operations to synchronize with Neptune. Do not kill this process.
[neptune] [info   ] All 5 operations synced, thanks for waiting!
[neptune] [info   ] Explore the metadata in the Neptune app: https://app.neptune.ai/kulikovka/AlexLenNets/e/AL-26/metadata


**LeNet5:** train acc-**0.79505** vs test acc-**0.7368** train loss-**0.443** vs test loss-**0.54*

**AlexNet5:** train acc-**0.9300** vs test acc-**0.9062** train loss-**0.17** vs test loss-**0.22*

**AlexNet5_stilled:** train acc-**0.9723** vs test acc-**0.9498** train loss-**0.07** vs test loss-**0.13*

### ***AlexNet5_stilled*** показала лучшие метрики и понадобилось всего **5 эпох** vs **15** у ***LeNet5*** и ***AlexNet5***.